### PR TITLE
Create models

### LINKED ISSUE #32: Create models
Create Classes for each entity following the messaging protocol.

### PR DESCRIPTION
**Pull request by celinecamacho**
Closes #32 
Models for the objects in the messaging protocol.

### CODE REVIEW

File: app/src/main/java/com/github/dedis/student20_pop/model/Event.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    private String name;
+    private long time;
+    private String id;
+    private String lao;
+    private List<String> attendees;
+    // Can use GeoLocation in the future
+    private String location;
+    // Can use enums in the future
+    private String type;
+    private JSONObject other;
```
Comment (GabrielFleischer) [inline app/src/main/java/com/github/dedis/student20_pop/model/Event.java]: What is that used for ?

File: app/src/main/java/com/github/dedis/student20_pop/model/Lao.java [hunk 2]
```diff
@@ -1,37 +1,173 @@
...
+     * @param events list of public keys of events, can be empty
+     * @throws IllegalArgumentException if the list is null or at least one public key is null
+     */
+    public void setEvents(List<String> events) {
+        if(events == null || events.contains(null)) {
+            throw new IllegalArgumentException("Trying to add a null event to the LAO " + name);
+        }
+        this.events = events;
+    }
+
```
Comment (jbmichel14) [inline app/src/main/java/com/github/dedis/student20_pop/model/Lao.java:153]: Would a addEvent method also be useful?

File: app/src/main/java/com/github/dedis/student20_pop/model/Keys.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+import java.util.Objects;
+
+/**
+ * Class modeling a set of Public and Private Keys
+ */
+public class Keys {
+
+    private String publicKey;
+    private String privateKey;
```
Comment (haoqianzhang) [inline app/src/main/java/com/github/dedis/student20_pop/model/Keys.java:11]: You may directly use the library in the future.

**Review by jbmichel14 — APPROVED** [review]

**Review by haoqianzhang — APPROVED** [review]
Great jobs! You may use the crypto library to support more operations.

File: app/src/main/java/com/github/dedis/student20_pop/model/Keys.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
+package com.github.dedis.student20_pop.model;
+
+import java.util.Objects;
+
+/**
+ * Class modeling a set of Public and Private Keys
+ */
+public class Keys {
```
Comment (GabrielFleischer) [inline app/src/main/java/com/github/dedis/student20_pop/model/Keys.java]: Maybe make it immutable

File: app/src/main/java/com/github/dedis/student20_pop/model/Election.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Class modeling an Election
+ */
+public class Election {
```
Comment (GabrielFleischer) [inline app/src/main/java/com/github/dedis/student20_pop/model/Election.java]: Maybe make it immutable

File: app/src/main/java/com/github/dedis/student20_pop/model/Event.java [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    private List<String> attestation;
```
Comment (GabrielFleischer) [inline app/src/main/java/com/github/dedis/student20_pop/model/Event.java]: Could this be a Set ?

Does the order matter ?

File: app/src/main/java/com/github/dedis/student20_pop/model/Event.java [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+     *
+     * @param attendees list of public keys of attendees, can be empty
+     * @throws IllegalArgumentException if the list is null or at least one public key is null
+     */
+    public void setAttendees(List<String> attendees) {
+        if(attendees == null || attendees.contains(null)) {
+            throw new IllegalArgumentException("Trying to add a null attendee to the event " + name);
+        }
+        this.attendees = attendees;
+    }
```
Comment (GabrielFleischer) [inline app/src/main/java/com/github/dedis/student20_pop/model/Event.java:115]: Is this really needed ? Aren't events one time objects with a fixed set of attendees at the creation ?

File: app/src/main/java/com/github/dedis/student20_pop/model/Event.java [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+/**
+ * Class modeling an Event
+ */
+public class Event {
+
+    private String name;
+    private long time;
+    private String id;
+    private String lao;
+    private List<String> attendees;
```
Comment (GabrielFleischer) [inline app/src/main/java/com/github/dedis/student20_pop/model/Event.java:20]: Could this be a Set ?

File: app/src/main/java/com/github/dedis/student20_pop/model/Election.java [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+
+    private String name;
+    private long time;
+    private String id;
+    private String lao;
+    private List<String> options;
+    private List<String> attestation;
```
Comment (GabrielFleischer) [inline app/src/main/java/com/github/dedis/student20_pop/model/Election.java]: Could those be Sets ?

File: app/src/main/java/com/github/dedis/student20_pop/model/Person.java [hunk 10]
```diff
@@ -0,0 +1,91 @@
+package com.github.dedis.student20_pop.model;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Class modeling a Person
+ */
+public class Person {
```
Comment (GabrielFleischer) [inline app/src/main/java/com/github/dedis/student20_pop/model/Person.java]: Should this be final ? At least for now ?

File: app/src/main/java/com/github/dedis/student20_pop/model/Person.java [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+
+    private String name;
+    private String id;
+    private String authentication;
+    private List<String> laos; // list of LAOs ids for now
```
Comment (GabrielFleischer) [inline app/src/main/java/com/github/dedis/student20_pop/model/Person.java]: Maybe set this as a set

File: app/src/main/java/com/github/dedis/student20_pop/model/Event.java [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Date;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Class modeling an Event
+ */
+public class Event {
```
Comment (GabrielFleischer) [inline app/src/main/java/com/github/dedis/student20_pop/model/Event.java]: Maybe set this as final. Or even immutable

File: app/src/main/java/com/github/dedis/student20_pop/model/Vote.java [hunk 13]
```diff
@@ -0,0 +1,80 @@
+package com.github.dedis.student20_pop.model;
+
+import java.util.Objects;
+
+/**
+ * Class modeling a Vote
+ */
+public class Vote {
```
Comment (GabrielFleischer) [inline app/src/main/java/com/github/dedis/student20_pop/model/Vote.java]: Set it as immutable ?

**Review by GabrielFleischer — CHANGES_REQUESTED** [review]
Some classes should be immutable, other should be final (at least for now)
Not sure about the massive usage of lists over sets

**Review by GabrielFleischer — COMMENTED** [review]
Maybe switch from Lists to Sets and get truly immutable obects for Event and Election

**Review by GabrielFleischer — APPROVED** [review]

File: app/src/main/java/com/github/dedis/student20_pop/model/Lao.java (hunk 2 above)
Comment (celinecamacho) [reply app/src/main/java/com/github/dedis/student20_pop/model/Lao.java:180]: The method setEvents can be used for this

File: app/src/main/java/com/github/dedis/student20_pop/model/Event.java (hunk 1 above)
Comment (celinecamacho) [reply app/src/main/java/com/github/dedis/student20_pop/model/Event.java]: It appears in the Protocol Messaging Definition, not sure of the usage yet.

File: app/src/main/java/com/github/dedis/student20_pop/model/Keys.java (hunk 4 above)
Comment (celinecamacho) [reply app/src/main/java/com/github/dedis/student20_pop/model/Keys.java]: It is now

File: app/src/main/java/com/github/dedis/student20_pop/model/Person.java (hunk 10 above)
Comment (celinecamacho) [reply app/src/main/java/com/github/dedis/student20_pop/model/Person.java]: It is now

File: app/src/main/java/com/github/dedis/student20_pop/model/Event.java (hunk 7 above)
Comment (celinecamacho) [reply app/src/main/java/com/github/dedis/student20_pop/model/Event.java:115]: The way I understood it, there's a possibility to create future events which don't have an attendees list/set yet

**Pull request merged**